diffusion/models/autoencoder.py

```diff
@@ -0,0 +1,613 @@
+    """Composer wrapper for the AutoEncoder.
+
+    Args:
+        model (AutoEncoder): AutoEncoder model to train.
+        autoencoder_loss (AutoEncoderLoss): Auto encoder loss module.
+        input_key (str): Key for the input to the model. Default: `image`.
+    """
+
+    def __init__(self, model: AutoEncoder, autoencoder_loss: AutoEncoderLoss, input_key: str = 'image'):
+        super().__init__()
+        self.model = model
+        self.autoencoder_loss = autoencoder_loss
+        self.input_key = input_key
+
+        # Set up train metrics
+        train_metrics = [MeanSquaredError()]
+        self.train_metrics = {metric.__class__.__name__: metric for metric in train_metrics}
+        # Set up val metrics
+        psnr_metric = PeakSignalNoiseRatio(data_range=2.0)
+        ssim_metric = StructuralSimilarityIndexMeasure(data_range=2.0)
+        lpips_metric = LearnedPerceptualImagePatchSimilarity(net_type='vgg')
+        val_metrics = [MeanSquaredError(), MeanMetric(), lpips_metric, psnr_metric, ssim_metric]
+        self.val_metrics = {metric.__class__.__name__: metric for metric in val_metrics}
+
+    def get_last_layer_weight(self) -> torch.Tensor:
+        """Get the weight of the last layer of the decoder."""
+        return self.model.get_last_layer_weight()
+
+    def forward(self, batch):
+        outputs = self.model(batch[self.input_key])
+        return outputs
+
+    def loss(self, outputs, batch):
+        last_layer = self.get_last_layer_weight()
+        return self.autoencoder_loss(outputs, batch, last_layer)
+
+    def eval_forward(self, batch, outputs=None):
+        if outputs is not None:
+            return outputs
+        outputs = self.forward(batch)
+        return outputs
+
+    def get_metrics(self, is_train: bool = False):
+        if is_train:
+            metrics = self.train_metrics
+        else:
+            metrics = self.val_metrics
+
+        if isinstance(metrics, Metric):
+            metrics_dict = {metrics.__class__.__name__: metrics}
+        elif isinstance(metrics, list):
+            metrics_dict = {metrics.__class__.__name__: metric for metric in metrics}
+        else:
+            metrics_dict = {}
+            for name, metric in metrics.items():
+                assert isinstance(metric, Metric)
+                metrics_dict[name] = metric
+
+        return metrics_dict
+
+    def update_metric(self, batch, outputs, metric):
+        clamped_imgs = outputs['x_recon'].clamp(-1, 1)
+        if isinstance(metric, MeanMetric):
+            metric.update(torch.square(outputs['latents']))
+        elif isinstance(metric, LearnedPerceptualImagePatchSimilarity):
+            metric.update(clamped_imgs, batch[self.input_key])
+        elif isinstance(metric, PeakSignalNoiseRatio):
+            metric.update(clamped_imgs, batch[self.input_key])
+        elif isinstance(metric, StructuralSimilarityIndexMeasure):
+            metric.update(clamped_imgs, batch[self.input_key])
+        else:
```

Comment:
Do you want to add an isintance assert about what the metric is suppose to be here?

Reply:
What would we assert it be? A generic `torchmetrics.Metric` perhaps? The `else` here follows the [composer classifier](https://github.com/mosaicml/composer/blob/f65139c20da1476a610294ed5bb76f2a15bbae0d/composer/models/tasks/classification.py#L122) to avoid enumerating the specific types of metrics that might be used. Although there are cases where this could break if we make the metrics configurable (they aren't ATM).

Reply:
If I understand, there is a fixed set of 5 metrics defined in the `__init__`. 4 of the metrics are caught by the above if-statement and the fifth one (`MeanSquaredError`, right?) is caught by this if-statement. It seems safer and consistent to check `elif isinstance(metric, MeanSquaredError)`

Reply:
Done!